src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -112,10 +114,40 @@ ErrorIfRelationIsAKnownShard(Oid relationId)
 	}
 
 	const char *relationName = get_rel_name(relationId);
+
 	ereport(ERROR, (errmsg("relation \"%s\" is a shard relation ", relationName)));
 }
 
 
+/*
+ * ErrorIfIllegallyChangingKnownShard errors out if the relation with relationId is
+ * a known shard and manual changes on known shards are disabled. This is
+ * valid for only non-citus (external) connections.
+ */
+void
+ErrorIfIllegallyChangingKnownShard(Oid relationId)
+{
+	if (GetCurrentLocalExecutionStatus() == LOCAL_EXECUTION_REQUIRED || 
+		IsCitusInitiatedRemoteBackend() || EnableManualChangesToShards)
+	{
+		return;
+	}
+
+	/* search the relation in all schemas */
+	bool onlySearchPath = true;
```

Comment:
I was wrong earlier, onlySearchPath should be false. I misinterpreted what it means. 

If it is true, we only consider a relation a known shard if it is on the search path, which is not what we want here.

I also kind of feel like it should not be an argument to RelationIsAKnownShard since it appears to have nothing to do with whether the relation is a known shard. Instead callers should check RelationIsVisible if they care.

Reply:
ping

Reply:
ping